From e355c0f338e221048e1e91bd9fb99e938049b37f Mon Sep 17 00:00:00 2001
From: Shreenidhi Shedi <sshedi@vmware.com>
Date: Sat, 8 Jan 2022 02:31:35 +0530
Subject: [PATCH] OpenSSLCryptoHash.cpp: Allow use of md5 in fips mode

Do this properly by comparing version of openssl in the system

Signed-off-by: Shreenidhi Shedi <sshedi@vmware.com>
---
 xsec/enc/OpenSSL/OpenSSLCryptoHash.cpp | 20 ++++++++++++++------
 xsec/enc/OpenSSL/OpenSSLCryptoHash.hpp |  8 +++++++-
 2 files changed, 21 insertions(+), 7 deletions(-)

diff --git a/xsec/enc/OpenSSL/OpenSSLCryptoHash.cpp b/xsec/enc/OpenSSL/OpenSSLCryptoHash.cpp
index 4087ce4..ad87fb9 100644
--- a/xsec/enc/OpenSSL/OpenSSLCryptoHash.cpp
+++ b/xsec/enc/OpenSSL/OpenSSLCryptoHash.cpp
@@ -58,8 +58,15 @@ OpenSSLCryptoHash::OpenSSLCryptoHash(HashType alg) :
         break;
 
     case (XSECCryptoHash::HASH_MD5) :
-    
+#if OPENSSL_VERSION_NUMBER >= 0x30000000
+        octx = OSSL_LIB_CTX_new();
+        mp_md = EVP_MD_fetch(octx, OSSL_DIGEST_NAME_MD5, "-fips");
+#else
+#ifdef EVP_MD_CTX_FLAG_NON_FIPS_ALLOW
+        EVP_MD_CTX_set_flags(mp_mdctx, EVP_MD_CTX_FLAG_NON_FIPS_ALLOW);
+#endif
         mp_md = EVP_get_digestbyname("MD5");
+#endif 
         break;
 
     case (XSECCryptoHash::HASH_SHA224) :
@@ -116,21 +123,22 @@ OpenSSLCryptoHash::OpenSSLCryptoHash(HashType alg) :
 
     EVP_DigestInit(mp_mdctx, mp_md);
     m_hashType = alg;
-
 }
 
-
 OpenSSLCryptoHash::~OpenSSLCryptoHash() {
+    EVP_MD_free((EVP_MD*)mp_md);
 #if (OPENSSL_VERSION_NUMBER < 0x10100000L)
     EVP_MD_CTX_cleanup(mp_mdctx);
 #else
     EVP_MD_CTX_free(mp_mdctx);
+#if OPENSSL_VERSION_NUMBER >= 0x30000000
+    if (m_hashType == HASH_MD5) {
+	    OSSL_LIB_CTX_free(octx);
+    }
+#endif
 #endif
-
 }
 
-
-
 // Hashing Activities
 void OpenSSLCryptoHash::reset(void) {
 
diff --git a/xsec/enc/OpenSSL/OpenSSLCryptoHash.hpp b/xsec/enc/OpenSSL/OpenSSLCryptoHash.hpp
index e5cb98f..48ff80e 100644
--- a/xsec/enc/OpenSSL/OpenSSLCryptoHash.hpp
+++ b/xsec/enc/OpenSSL/OpenSSLCryptoHash.hpp
@@ -40,6 +40,10 @@
 
 #   include <openssl/evp.h>
 
+#if OPENSSL_VERSION_NUMBER >= 0x30000000
+#   include <openssl/core_names.h>
+#endif
+
 /**
  * @ingroup opensslcrypto
  */
@@ -156,7 +160,9 @@ private:
     unsigned char       m_mdValue[EVP_MAX_MD_SIZE];     // Final output
     unsigned int        m_mdLen;                        // Length of digest
     HashType            m_hashType;                     // What type of hash is this?
-
+#if OPENSSL_VERSION_NUMBER >= 0x30000000
+    OSSL_LIB_CTX 	*octx;
+#endif
 };
 
 #endif /* XSEC_HAVE_OPENSSL */
-- 
2.30.0

